07_ansible_operator/README.org

```diff
@@ -34,3 +41,220 @@ Customize options such as username/password domain in ~nextcloud.yml~ according
 #+begin_src bash
 doctl compute droplet delete nextcloud 
 #+end_src
+
+** Create operator which deploys nextcloud
+
+WE are lazy here so we decided to go with nextcloud image from DockerHub. In real life scenario
+we have to be much more curious and we should setup a security GW to check the 3rd party images
+if theyre good enough to be used in our environment.
+
+Now its time to jump into operator waters. We will generate our NextCloud Ansible operator
+directory structure by executing:
+#+BEGIN_SRC bash
+operator-sdk new nextcloud --type ansible --kind Nextcloud --api-version cz.prgcont/v1alpha1 --skip-git-init
+#+END_SRC
+
+This should end up with a directory structure like:
+#+BEGIN_SRC 
+nextcloud/
+├── build
+│   ├── Dockerfile
+│   └── test-framework
+│       ├── ansible-test.sh
+│       └── Dockerfile
+├── deploy
+│   ├── crds
+│   │   ├── cz_v1alpha1_nextcloud_crd.yaml
+│   │   └── cz_v1alpha1_nextcloud_cr.yaml
+│   ├── operator.yaml
+│   ├── role_binding.yaml
+│   ├── role.yaml
+│   └── service_account.yaml
+├── molecule
+│   ├── default
+│   │   ├── asserts.yml
+│   │   ├── molecule.yml
+│   │   ├── playbook.yml
+│   │   └── prepare.yml
+│   ├── test-cluster
+│   │   ├── molecule.yml
+│   │   └── playbook.yml
+│   └── test-local
+│       ├── molecule.yml
+│       ├── playbook.yml
+│       └── prepare.yml
+├── roles
+│   └── nextcloud
+│       ├── defaults
+│       │   └── main.yml
+│       ├── files
+│       ├── handlers
+│       │   └── main.yml
+│       ├── meta
+│       │   └── main.yml
+│       ├── README.md
+│       ├── tasks
+│       │   └── main.yml
+│       ├── templates
+│       └── vars
+│           └── main.yml
+└── watches.yaml
+#+END_SRC
+
+The most important file for us is *watches.yaml*. It defines mapping between CR and Ansible Role. 
+This means that everytime there is a change in a watched Kubernetes object the Ansible role specifed 
+in *watches.yaml* file will be executed.
+
+*Note:* As you see your role is executed frequently so it really must be idempotent.
+
+For more real world style operators please also read about [[https://github.com/operator-framework/operator-sdk/blob/master/doc/ansible/dev/finalizers.md][finalizers]], which enables you to clear
+all resources which are not stored in Kubernetes.
+
+
+** Test our operator skelet
+To start our development cycle we need to add *CR* definition into our Kubernetes cluster. It can 
+be acchieved by executing:
+
+#+BEGIN_SRC bash
+kubectl create -f deploy/crds/cz_v1alpha1_nextcloud_cr.yaml
```

Comment:
Shouldn't this be step 2, first you add API using `cz_v1alpha1_nextcloud_crd.yaml`, then create Custom Resource instance: `cz_v1alpha1_nextcloud_cr.yaml`